### ISSUE TITLE
Define what confidence means, and record that it is not comparable across sources

### ISSUE THREAD

**Issue by roaldarbol** [animovement/anicore#150]
### What problem would this solve?

`confidence` is produced by eight `aniread` readers and consumed by `anicheck::check_confidence()` and `aniprocess::filter_na_across(method = "confidence")`, but it is **defined nowhere**. `vignettes/articles/aniframe-structure.qmd` documents what / when / where, the index and coordinate systems, and never mentions it. Neither does `list_default_metadata()`, nor any `aniframe` help page.

That absence has two consequences that are already real, not hypothetical.

**1. The values are not comparable across sources, and nothing says so.**

Each reader fills the column from whatever its format happens to carry:

| Reader | `confidence` comes from |
|---|---|
| `read_deeplabcut()`, `read_lightningpose()` | `likelihood` |
| `read_anipose()` | `score` — documented as "mean detection confidence across cameras (0-1 scale)" |
| `read_idtracker()` | identity assignment probability |
| `read_sleap()` | `point_scores` |
| `read_freemocap()` | `1 / (1 + reprojection_error)` (#117) |
| `read_bonsai()`, `read_animalta()`, `read_trex()` | `NA` — the format carries none |

A DeepLabCut likelihood of 0.5 and a FreeMoCap 0.5 mean different things. Thresholding is where this bites: someone picks `filter_na_across(threshold = 0.5)` from DeepLabCut intuition and applies it to another source. The ordering within one recording is always meaningful; the absolute value across sources is not.

**2. SLEAP confidences exceed 1, and nothing handles it.**

Not an edge case. Measured on `SLEAP_two-mice_octagon` (the movement netCDF in `animovement/movement-data`, exported from the GIN SLEAP file):

```
n     = 123,840
min   = 0.200
median= 0.907
max   = 1.467
> 1   = 24,571  (19.8%)
```

Roughly a fifth of the values are above 1. `read_sleap()` reads `point_scores` straight through with no clamping, so any code assuming a `(0, 1]` range is wrong on SLEAP data today. `anicheck::check_confidence()` draws violins that will run off the end of a `[0, 1]` axis, and a "keep the top 10%" heuristic behaves differently than expected.

@roaldarbol recalls Talmo Pereira advising that anything above 1 be treated as 1. I could not corroborate that in writing: the `movement` Python package passes SLEAP confidence through unchanged (`movement/io/load_poses.py`), and I found nothing in the SLEAP docs or issue tracker stating it. **So the clamping question is open and worth confirming with the SLEAP developers before acting on it** — this issue records the measurement, not a decision.

### What would you like to happen?

**Define `confidence` in `aniframe-structure.qmd`, as the canonical home.** anicore owns the data model and that article already describes the standard columns. Four points:

- per-observation, aligned with the row it sits on
- higher is better
- conventionally, but not guaranteed, in `(0, 1]`
- **source-dependent in meaning**: comparable within a recording, not across sources

Everything else should link here rather than restate it.

Then, in the packages that touch it (separate issues, referencing this one):

- **aniread** — each reader's `@return` says what its confidence actually is. `read_anipose()` and `read_freemocap()` already do; most only say the column exists.
- **aniprocess** — one line on `filter_na_across(method = "confidence")`, because choosing a threshold is where the mistake happens.
- **anicheck** — `check_confidence()` should not assume a `[0, 1]` axis, and should say that comparing distributions across sources is not meaningful.

**Optionally, make it machine-readable** rather than prose: a `confidence` column on `aniread::get_supported_sources()` naming the underlying quantity per source (`"likelihood"`, `"score"`, `"probability"`, `"1/(1+reprojection_error)"`, `NA`). That turns "we wrote it down" into something a test can assert, and pairs with `source_format` — both answer "what exactly am I holding?".

### What alternatives have you considered?

- **Normalise everything to `(0, 1]` at read time.** Tempting, and it would make thresholds portable. But it discards information, hides the SLEAP question rather than answering it, and a normalised DeepLabCut likelihood still would not mean the same as a normalised reprojection error. Ordinal comparability within a source is all that is actually available.
- **Put the note only in `read_freemocap()`.** Where it currently lives. It reads as a FreeMoCap quirk when it is a property of the column, and it leaves the SLEAP range problem undocumented.
- **A separate `confidence_type` metadata field.** Heavier than needed while nothing consumes it; the `get_supported_sources()` column gives most of the value.
- **Do nothing.** The column has worked so far. But `read_freemocap()` has just added a source whose confidence is a transformed error, which makes the ambiguity harder to ignore.

### Anything else?

The measurement above is reproducible from `animovement/movement-data`:

```sh
ncdump -v confidence data/movement/SLEAP_two-mice_octagon.analysis-1768334869096.nc
```

Related: animovement/aniread#117 (the FreeMoCap mapping that prompted this), animovement/aniread#118 (the reader-side `format` convention, which has the same "say what you are holding" shape).

**Comment by roaldarbol** [comment]
Traced where SLEAP's values actually come from, which explains the >1 range and sharpens what this issue is really about.

## Where SLEAP's `point_scores` come from

Not something SLEAP invented — it is the heatmap-regression formulation inherited from human pose estimation (Tompson 2014; Stacked Hourglass, Newell 2016; OpenPose/PAFs, Cao 2017 — SLEAP's `PartAffinityFieldsHead` is the Cao construct directly).

**1. The training target is a Gaussian of amplitude exactly 1** (`sleap_nn/data/confidence_maps.py`):

```python
cm = torch.exp(-((xv - x)**2 + (yv - y)**2) / (2 * sigma**2))
```

**2. The output head is linear with MSE loss** (`sleap_nn/architectures/heads.py`). The base `Head` returns `activation = "identity"`, `loss_function = "mse"`, and none of the four confidence-map heads override it:

| Head | Activation |
|---|---|
| `SingleInstanceConfmapsHead` | inherits `identity` |
| `CentroidConfmapsHead` | inherits `identity` |
| `CenteredInstanceConfmapsHead` | inherits `identity` |
| `MultiInstanceConfmapsHead` | inherits `identity` |
| `ClassMapsHead` | **`sigmoid`** |
| `ClassVectorsHead` | **`softmax`** |

The identity/classification heads *are* bounded. The confidence maps deliberately are not.

**3. The score is the raw tensor value at the peak pixel** (`sleap_nn/inference/ops/peaks.py:214`):

```python
peak_vals = cms[peak_subs[:, 0], peak_subs[:, 3], peak_subs[:, 1], peak_subs[:, 2]]
```

Subpixel refinement (`integral_regression`) refines the coordinates and returns `peak_vals` untouched.

So `point_scores` is **an unbounded linear regression output approximating a Gaussian of height 1**. A value of 1.47 is the network overshooting the target amplitude — routine for MSE regression with a linear output. It is not a probability and was never meant to be one.

DeepLabCut descends from DeeperCut, which puts a **sigmoid** on its part-detection scoremap with cross-entropy loss. That is exactly why DLC likelihoods are bounded and SLEAP's are not: same-looking column, opposite output layers.

This also reframes the clamping question. Clamping SLEAP at 1 is not a normalisation — it discards the overshoot that distinguishes a very confident detection from a merely confident one, across ~20% of the data. If comparability is the aim, dividing by the per-recording max or rank-transforming preserves ordering; clamping flattens the top fifth into a tie.

## What we are squashing together

| Source | Underlying quantity | Bounded? | What it actually means |
|---|---|---|---|
| SLEAP | Confidence-map peak value | **No** (linear + MSE) | How closely the network reproduced a unit-height Gaussian |
| DeepLabCut / LightningPose | `likelihood` | `[0,1]` (sigmoid + CE) | Per-pixel probability that this is the part |
| Anipose | `score` | `[0,1]` | **Mean** of the 2D detection scores across cameras |
| idtracker.ai | Identity probability | `[0,1]` | Probability this blob is **individual _k_** |
| FreeMoCap | `1/(1+reprojection_error)` | `(0,1]` | Multi-view geometric consistency — no network involved |
| Bonsai, AnimalTA, TRex | — | — | `NA`; centroid trackers with no per-point score |

Worth being precise about FreeMoCap, because it is easy to misread: reprojection error is the pixel distance between the 2D detections and the reprojection of the triangulated 3D point. It measures whether the cameras **agree with each other**, not whether they are right. If every camera confidently puts the wrist in the same wrong place, the error is near zero.

Anipose is the useful precedent here: it exports `_score`, `_error` **and** `_ncams` as separate columns per keypoint, rather than choosing between detector certainty and geometric consistency. They answer different questions.

## The part that needs more thought

Three genuinely different things share this column, and the differences line up with an axis anicore already has:

| Certainty about | Sources |
|---|---|
| **where** — localisation | SLEAP, DLC/LP, anipose score, FreeMoCap reprojection |
| **what** — identity | idtracker.ai probability, TRex categories |

So `confidence` is not one column that varies in scale; it is **two roles wearing one name**, and `variables_what` / `variables_where` is already the vocabulary for that distinction. Something like `confidence_where` / `confidence_what` follows naturally.

But @roaldarbol raised an objection that stops this being a straightforward rename, and it is worth recording before anyone implements it:

> it may be correct at import, but if the user then adds an "individual" what variable, then it also sounds like the confidence was about attribution to the correct individual, not just keypoint

That is right, and it generalises. A slot-keyed name is only stable while the slot is. A DLC likelihood imported into a single-animal frame is certainty about *keypoint localisation*; add an `individual` column to `variables_what` and `confidence_what` starts implying certainty about identity assignment, which that number never spoke to. The name would silently become a claim the data does not support.

The underlying problem is that certainty attaches to **a specific claim**, not to a slot — "this point is at (x, y)", "this keypoint is the left ear", "this blob is individual 3" — and slots can gain columns long after import. Naming by slot fixes the scale problem and introduces a provenance problem.

Directions, none of them decided:

- A certainty column names **which column(s) it qualifies**, rather than which slot — stable under later edits, at the cost of being wordier.
- Certainty carries its own metadata (quantity, bounds, what it qualifies), with `confidence` staying a plain column.
- Keep one `confidence`, and put the role in metadata rather than the column name.

Flagging rather than proposing: this needs design work, and the naming is the easy half.

**Comment by roaldarbol** [comment]
A fourth kind of certainty, found while adding TRex's `.npz` reader (animovement/aniread#120).

## TRex does have a confidence — in its other export

The table above lists TRex under "no per-point score". That is true of its CSV export and false of the software. TRex's native `.npz` export carries a `detection_p` array, one value per frame per individual, in `[0, 1]`. On the five-locusts recording it runs 0.21–0.95.

So `read_trex()` returns all-`NA` confidence from one export and real values from the other, for the same recording. `source_format` distinguishes them, which is exactly the case #118 and the `source_format` field were added for — but it means a single source name maps to two different confidence semantics, one of which is "absent".

## And it is a third kind of claim

`detection_p` is the probability that a **blob is an animal at all** — a detection, before any question of where its parts are or which individual it belongs to. That is neither of the two roles in the table above:

| Certainty about | Sources |
|---|---|
| **detection** — is there an animal here? | TRex `detection_p` |
| **where** — localisation | SLEAP, DLC/LP, anipose score, FreeMoCap reprojection |
| **what** — identity | idtracker.ai probability |

TRex is interesting because it separates them internally: `detection_p` for the detection, and a `missing` flag for frames where there is no detection at all. Its identity confidence, from the visual-identification step, is a further thing again and is not in the per-individual arrays.

So the column now holds four different claims, not three. That does not change the shape of the problem, but it does weaken one argument: if a naming scheme has to accommodate detection, localisation and identity, `confidence_where` / `confidence_what` is already one short, and the next reader may well bring a fifth.

## Which makes @roaldarbol's objection harder to design around, not easier

The objection above was that a slot-keyed name is only stable while the slot is — add an `individual` column later and `confidence_what` starts implying something the number never said. TRex adds a second failure mode from the other direction: `detection_p` qualifies *no* slot. There is no `variables_detection`, and a detection is not an identity or a position; it is a prior condition for both.

That is some evidence for the third direction listed above — certainty carrying its own metadata (what quantity, what bounds, what claim it makes) rather than the role being encoded in a column name at all. A name has to be chosen from a closed set decided now; a metadata field can describe a claim no one has thought of yet.

Still flagging rather than proposing.

## Practical note

`animovement/movement-data`'s `metadata.yaml` records TRex as having no confidence. That is correct for the `beetle.csv` fixture it describes and wrong as a statement about TRex, so that entry needs splitting by export once animovement/aniread#120 lands.

**Comment by roaldarbol** [comment]
SLEAP is worse than the table above says. It does not carry one confidence — it carries **five**, distinguished in its own data model, and `read_sleap()` returns one of them.

## The five, in sleap-io's own words

From the `Instance` / `PredictedInstance` docstrings in `sleap_io/model/instance.py`:

| Field | What sleap-io says it is |
|---|---|
| per-point score | Per-keypoint confidence, on `PredictedPointsArray` |
| `score` | "The instance **detection or part grouping** prediction score… the confidence with which this entire instance was predicted" |
| `tracking_score` | "The score associated with the `Track` assignment… typically the value from the score matrix used in an identity assignment" — a *short-term tracklet* |
| `identity_score` | "The score associated with the `identity` assignment… Kept separate from `tracking_score` (short-term tracklet vs long-term identity)" |
| `category_score` | "The score associated with the `category` assignment" — the *class* an instance belongs to |

The analysis exports surface three of them: the h5 has `point_scores`, `instance_scores` and `tracking_scores`; the CSV has `{node}.score`, `instance.score` and `track_score`.

`read_sleap()` takes the per-point score and drops the rest, in both exports (animovement/aniread#124). That is the right choice for a single `confidence` column, and it is lossy in a way nothing records.

## Two of them, in one file, disagree about their range

Measured on `SLEAP_three-mice_Aeon_mixed-labels.analysis.h5`:

```
point_scores      n=1558   range=[0.5019, 1.1596]   500 values > 1
instance_scores   n=1558   range=[0.4038, 0.9999]     0 values > 1
tracking_scores   n=1558   range=[0, 0]               0 values > 1
```

Nearly a third of the per-point scores exceed 1, and none of the instance scores do. That is not noise: it follows from the earlier trace up-thread. The confidence-map heads are linear with MSE loss, so their peak values are unbounded; the instance score comes out of grouping and is bounded. Two numbers, both called "score", in one file, on different scales, for different claims.

`tracking_scores` being all zero is its own kind of trap. sleap-io says the score is `None` "if the track was assigned manually", and this file is the *mixed-labels* fixture — part predicted, part human-corrected. So a zero here means "not applicable", not "no confidence". Anything treating it as a number would rank every manually-verified track as the least trustworthy in the file, which is exactly backwards.

## Why this matters for the naming question

The taxonomy in the comment above had detection, where, and what. SLEAP alone fills four of those boxes and splits one:

- per-point score → **where**
- `score` → **detection**, or part grouping, which is arguably neither detection nor localisation but "do these parts belong to one animal"
- `tracking_score` → **what**, short-term
- `identity_score` → **what**, long-term — and sleap-io keeps these two deliberately separate
- `category_score` → the class, which is a fifth thing again

So `confidence_what` would have to serve both `tracking_score` and `identity_score`, which the upstream format explicitly refuses to conflate. A naming scheme fixed now has to guess how many kinds there will be; this is the third unanticipated kind in as many readers, and the first where one source distinguishes two flavours of the *same* role.

That is further evidence for the third direction listed earlier — certainty carrying its own metadata (what quantity, what bounds, what claim, whether it is even applicable) rather than the role being encoded in a column name.

## Practical note

Whatever is decided, `read_sleap()` currently drops four scores silently. Even without a design, it would be worth the reader saying which one it kept — `source_format` records which export was read (animovement/aniread#124), but nothing records which of the file's five scores became `confidence`.
